Add Motoko ICRC2-Swap example

0xAegir opened the pull request:
**Overview**
Why do we need this feature? What are we trying to accomplish?

The goal of this specific example is to provide a minimalist canister example for new ecosystem developers wishing to safely handle deposits & withdrawals of ICRC-2 tokens. The code should be fully documented, and include a few tests, so the example can act as a learning and reference material.

**Requirements**
What requirements are necessary to consider this problem solved?

To avoid confusing new learners, I've kept the example as simple as possible. It handles three methods, “deposit”, “swap”, and “withdraw” (with a "balances" method added for debugging).

The "swap" is not strictly necessary, but it acts as a simple motivating use-case for the code to show how developers can work with with deposited balances.

The example also includes several test-cases. The purpose of these javascript tests is to show how a potential frontend would use the provided example canister.

This code will likely be copied throughout the ecosystem, so ideally this code would be heavily audited. However, due to cost, audits are out-of-scope. The code should, at least, undergo a thorough security-review from the Dfinity team. 🙏 

For now I'm only adding the example in Motoko (because that is where my expertise is), but adding a matching Rust example would be ideal.

sesi200 commented on motoko/icrc2-swap/LICENSE:
```diff
@@ -0,0 +1,7 @@
+Copyright 2023 0xAegir <0xAegir@protonmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, sub ...
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WI ...
```
Not sure I'm allowed to merge something that includes a separate license. Let me check...

0xAegir replied:
Sure, happy to re-assign the license to whatever you want. I just copied the one from the `ic-pos` example, since I assumed that was the norm.

sesi200 commented on motoko/icrc2-swap/canister_ids.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "swap": {
+    "local": "bkyz2-fmaaa-aaaaa-qaaaq-cai"
+  },
+  "token_a": {
+    "local": "bd3sg-teaaa-aaaaa-qaaba-cai"
+  },
+  "token_b": {
+    "local": "be2us-64aaa-aaaaa-qaabq-cai"
+  }
+}
```
This file only holds canister ids on 'persistent' networks like mainnet. As `local` is frequently wiped it is 'ephemeral' and therefore will ignore these canister ids. The entire file can be removed

sesi200 commented on motoko/icrc2-swap/README.md:
```diff
@@ -0,0 +1,311 @@
+- **Swap Tokens**: Users can swap the tokens for each other. This is implemented
+  in a very simple naive 1:1 manner. The point is just to demonstrate some
+  minimal behavior.
+- **Withdraw Tokens**: Users can send withdraw the resulting tokens after
```
```suggestion
- **Withdraw Tokens**: Users can withdraw the resulting tokens after
```
or did you mean `send/withdraw`?

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+Deploy Token A:
+
+```bash
+dfx deploy --network local token_a --argument '
````
```suggestion
dfx deploy token_a --argument '
```
implicit, can be skipped

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+Deploy Token B:
+
+```bash
+dfx deploy --network local token_b --argument '
````
```suggestion
dfx deploy token_b --argument '
```
same here

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+The swap canister accepts deposits, and performs the swap.
+
+```bash
+export TOKEN_A=$(dfx canister id --network local token_a)
+export TOKEN_B=$(dfx canister id --network local token_b)
+
+dfx deploy --network local swap --argument '
````
```suggestion
export TOKEN_A=$(dfx canister id token_a)
export TOKEN_B=$(dfx canister id token_b)

dfx deploy swap --argument '
```

sesi200 commented on motoko/icrc2-swap/README.md:
```diff
@@ -0,0 +1,311 @@
+  }
+'
+
+export SWAP=$(dfx canister id --network local swap)
```
```suggestion
export SWAP=$(dfx canister id swap)
```

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+```bash
+# Approve Bob to deposit 1.00000000 of Token B, and 0.0001 extra for the
+# transfer fee
+dfx canister --network local call --identity alice token_a icrc2_approve '
````
```suggestion
dfx canister call --identity alice token_a icrc2_approve '
```

sesi200 commented on motoko/icrc2-swap/README.md:
```diff
@@ -0,0 +1,311 @@
+
+# Approve Bob to deposit 1.00000000 of Token B, and 0.0001 extra for the
+# transfer fee
+dfx canister --network local call --identity bob token_b icrc2_approve '
```
```suggestion
dfx canister call --identity bob token_b icrc2_approve '
```

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+
+```bash
+# Deposit Alice's tokens
+dfx canister --network local call --identity alice swap deposit 'record {
````
```suggestion
dfx canister call --identity alice swap deposit 'record {
```

sesi200 commented on motoko/icrc2-swap/README.md:
```diff
@@ -0,0 +1,311 @@
+}'
+
+# Deposit Bob's tokens
+dfx canister --network local call --identity bob swap deposit 'record {
```
```suggestion
dfx canister call --identity bob swap deposit 'record {
```

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+'
+```
+
+Then we can perform the deposit to transfer the tokens from our wallet to the swap canister:
````
Do you think we could reword this? IMO this sounds like _we_ do the transfer when it is the `swap` canister that performs the transfer. While we two understand this easily, I think it will confuse people that are new to the whole ICRC-2 thing

0xAegir replied:
Good point. I tried rewriting this. What do you think?

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+## Step 5: Perform a swap
+
+```bash
+dfx canister --network local call swap swap 'record {
````
```suggestion
dfx canister call swap swap 'record {
```

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+We can check the deposited balances with:
+
+```bash
+dfx canister --network local call swap balances
````
```suggestion
dfx canister call swap balances
```

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+
+```bash
+# Withdraw Alice's Token B balance (1.00000000), minus the 0.0001 transfer fee
+dfx canister --network local call --identity alice swap withdraw 'record {
````
```suggestion
dfx canister call --identity alice swap withdraw 'record {
```

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+
+```bash
+# Withdraw Bob's Token A balance (1.00000000), minus the 0.0001 transfer fee
+dfx canister --network local call --identity bob swap withdraw 'record {
````
```suggestion
dfx canister call --identity bob swap withdraw 'record {
```

sesi200 commented on motoko/icrc2-swap/README.md:
````diff
@@ -0,0 +1,311 @@
+
+```bash
+# Check Alice's Token A balance. They should now have 998.99980000 A
+dfx canister --network local call token_a icrc1_balance_of 'record {
````
```suggestion
dfx canister call token_a icrc1_balance_of 'record {
```

sesi200 commented on motoko/icrc2-swap/README.md:
```diff
@@ -0,0 +1,311 @@
+}'
+
+# Check Bob's Token A balance, They should now have 0.99990000 A.
+dfx canister --network local call token_a icrc1_balance_of 'record {
```
```suggestion
dfx canister call token_a icrc1_balance_of 'record {
```

sesi200 commented on motoko/icrc2-swap/README.md:
```diff
@@ -0,0 +1,311 @@
+
+- Any DeFi on the Internet Computer is experimental. It is a constantly evolving
+  space, with unknown attacks, and should be treated as such.
+- Due to the nature of inter-canister messaging on the IC, it is possible for
```
```suggestion
- Due to the nature of asynchronous inter-canister messaging on the IC, it is possible for
```

sesi200 commented on motoko/icrc2-swap/README.md:
```diff
@@ -0,0 +1,311 @@
+# ICRC-2 Swap
```
Very well written! Thank you very much for putting in so much effort. I only have some minor comments

sesi200 commented on motoko/icrc2-swap/dfx.json:
```diff
@@ -0,0 +1,26 @@
+      "wasm": "https://download.dfinity.systems/ic/2e3589427cd9648d4edaebc1b96b5daf8fdd94d8/canisters/ic-icrc1-ledger.wasm.gz"
+    }
+  },
+  "defaults": {
+    "build": {
+      "args": "",
+      "packtool": ""
+    }
+  },
```
```suggestion
```
Can safely be removed

sesi200 commented on motoko/icrc2-swap/src/swap/main.mo:
```diff
@@ -0,0 +1,294 @@
+  private var balancesA = TrieMap.TrieMap<Principal, Nat>(Principal.equal, Principal.hash);
+  private var balancesB = TrieMap.TrieMap<Principal, Nat>(Principal.equal, Principal.hash);
+
+  // Because TrieMaps are not directly storeable in stable memory we need a
```
```suggestion
  // Because TrieMaps are not directly storable in stable memory we need a
```

sesi200 commented on motoko/icrc2-swap/src/swap/main.mo:
```diff
@@ -0,0 +1,294 @@
+    let block_height = switch (transfer_result) {
+      case (#Ok(block_height)) { block_height };
+      case (#Err(err)) {
+        // Transfer failed. There's no cleanup for us to do, so we can just
```
```suggestion
        // Transfer failed. There's no cleanup for us to do since no state has changed, so we can just
```
Since this is aimed at beginners I think it's worth being very explicit

sesi200 commented on motoko/icrc2-swap/src/swap/main.mo:
```diff
@@ -0,0 +1,294 @@
+      };
+    };
+
+    // From here on out, we need to make sure that this function does *not*
+    // fail. If it failed, the token transfer would be complete (meaning we
+    // would have the user's tokens), but we would not have credited their
+    // account yet, so this canister would not *know* that it had received the
+    // user's tokens.
+    // 
+    // If the function *can* fail here after this point, we should either:
+    // - Move that code to a separate action later
+    // - Have failure-handling code which returns the user's tokens
```
:heart: I'll point to this when someone asks me how to write example code

sesi200 commented on motoko/icrc2-swap/src/swap/main.mo:
```diff
@@ -0,0 +1,294 @@
+  // Swap TokenA for TokenB
+  // - Exchange tokens between the two given users.
+  // - For this example, there will be no clever swap mechanism, it simply swaps all
+  //   deposits for the two users. Designing a useful and safer swap mechanism
```
```suggestion
  //   deposits for the two users, even allowing anybody to perform the swap. Designing a useful and safer swap mechanism
```
Let's make it even more obvious that this is _not_ the way to do it :)

sesi200 commented on motoko/icrc2-swap/src/swap/main.mo:
```diff
@@ -0,0 +1,294 @@
+  public shared(msg) func swap(args: SwapArgs): async Result.Result<(), SwapError> {
+    // Because both tokens were deposited before calling swap, we can execute
+    // this function atomically. To do that there must be no `await` calls in
+    // this function. Additionally, we need to be careful with the order that
+    // we update any internal state variables. If this function were to update
+    // some state variables, then fail, before updating others, it could leave
+    // this canister with inconsistent internal state.
```
Wrong. This is only a concern if you `await` in the middle of state changes. If this code traps anywhere all state changes will be reverted

0xAegir replied:
Ah true. 🤔 Let me rewrite this...

sesi200 commented on motoko/icrc2-swap/src/swap/main.mo:
```diff
@@ -0,0 +1,294 @@
+  };
+
+  // Allow withdrawals
+  // - Allow UserA to withdraw TokenB, and UserB to withdraw TokenA
```
This function allows anyone to withdraw any token they hold
```suggestion
```

sesi200 commented on motoko/icrc2-swap/package-set.dhall:
```diff
@@ -0,0 +1,25 @@
+let upstream = https://github.com/dfinity/vessel-package-set/releases/download/mo-0.6.21-20220215/package-set.dhall sha256:b46f30e811fe5085741be01e126629c2a55d4c3d6ebf49408fb3b4a98e37589b
+let Package =
+    { name : Text, version : Text, repo : Text, dependencies : List Text }
+
+let
+  -- This is where you can add your own packages to the package-set
+  additions =
+    [] : List Package
+
+let
+  {- This is where you can override existing packages in the package-set
+
+     For example, if you wanted to use version `v2.0.0` of the foo library:
+     let overrides = [
+         { name = "foo"
+         , version = "v2.0.0"
+         , repo = "https://github.com/bar/foo"
+         , dependencies = [] : List Text
+         }
+     ]
+  -}
+  overrides =
+    [] : List Package
+
+in  upstream # additions # overrides
```
I don't see any code that actually uses vessel. Can this be removed?

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -0,0 +1,130 @@
+  (dfx canister info "$1" 2>&1 | grep 'Module hash: 0x')
+}
+
+# echo
+# echo == Dependencies.
+# echo
+
+# vessel install
```
```suggestion
```

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -0,0 +1,130 @@
+# The JS tests use a different principal, so we should give that some balance as well.
+export JS="jg6qm-uw64t-m6ppo-oluwn-ogr5j-dc5pm-lgy2p-eh6px-hebcd-5v73i-nqe"
+
+# echo
+# echo == Install NNS.
+# echo
+
+# if canister_exists "nns-governance"; then
+#     echo "nns-governance already exists skipping nns install" >&2
+# else
+#     dfx extension install nns
+#     dfx nns install --identity icrc2-swap-minter
+#     dfx ledger transfer --identity icrc2-swap-minter --memo 0 --amount 10000 $(dfx ledger account-id)
+# fi
```
```suggestion
```

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -0,0 +1,130 @@
+echo == Deploy: Token A
+echo
+
+dfx deploy --network local token_a --argument "
```
```suggestion
dfx deploy token_a --argument "
```

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -0,0 +1,130 @@
+echo == Deploy: Token A
+echo
+
+dfx deploy --network local token_b --argument "
```
```suggestion
dfx deploy token_b --argument "
```

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -0,0 +1,130 @@
+echo == Deploy: Swap
+echo
+
+dfx deploy --network local swap --argument "
```
```suggestion
dfx deploy swap --argument "
```

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -0,0 +1,130 @@
+  }
+"
+
+export SWAP=$(dfx canister id --network local swap)
```
```suggestion
export SWAP=$(dfx canister id swap)
```

sesi200 commented on motoko/icrc2-swap/test/swap/swap.test.ts:
```diff
@@ -0,0 +1,261 @@
+      const alice = newIdentity();
+      await fundIdentity(tokenA(minter), alice, 100_020_000n);
+
+      // Give bob, just enough B
```
```suggestion
      // Give bob just enough B
```

sesi200 commented on motoko/icrc2-swap/test/swap/swap.test.ts:
```diff
@@ -0,0 +1,261 @@
+      });
+      expect(approvalB).toHaveProperty("Ok");
+
+      // Alice deposits 1 A
```
```suggestion
      // Bob deposits 1 A
```

sesi200 commented on motoko/icrc2-swap/test/swap/swap.test.ts:
```diff
@@ -0,0 +1,261 @@
+        // Give alice just enough A
+        const alice = newIdentity();
+
+        // Alice tries to deposit 1 A. This will fail because only 0.5A has
+        // been approved.
```
```suggestion
        // Alice tries to deposit a token that does not exist
```

0xAegir replied:
Hmm.. I disagree on this one. The token canister exists, but in this scenario Alice hasn't approved any transfers.

0xAegir replied:
Oh, right, no I see. It got copied to this test.

0xAegir commented:
I added a .github action for the new CI job, but it probably needs approval.

I *think* I've addressed all the feedback. LMK if I've missed anything.

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -0,0 +1,112 @@
+if (dfx identity list | grep icrc2-swap-minter 2>&1 >/dev/null) ; then
+    echo "icrc2-swap-minter account already exists" >&2
+else
+    dfx identity import icrc2-swap-minter <(cat <<EOF
```
```suggestion
    dfx identity import icrc2-swap-minter -- storage-mode plaintext <(cat <<EOF
```
CI had a problem with passwords. This should fix it

sesi200 reviewed (changes requested):
Awesome, thank you for addressing all of this! CI now has two problems, once these are resolved we should be good to merge

* Keyring is not available on CI, so dfx defaults to password-protected PEM files. If we force plaintext then it should not get stuck there
* There is an error with `npm install` on Linux. Mind having a look?

sesi200 commented on motoko/icrc2-swap/test/deploy.sh:
```diff
@@ -13,7 +13,7 @@ echo
     echo "icrc2-swap-minter account already exists" >&2
 else
-    dfx identity import icrc2-swap-minter <(cat <<EOF
+    dfx identity import icrc2-swap-minter -- storage-mode plaintext <(cat <<EOF
```
```suggestion
    dfx identity import icrc2-swap-minter --storage-mode plaintext <(cat <<EOF
```
typo?

0xAegir replied:
Yes, just fixed

sesi200 commented on .github/workflows/motoko-icrc2-swap-example.yml:
```diff
@@ -0,0 +1,41 @@
+    runs-on: macos-12
+    steps:
+      - uses: actions/checkout@v1
+      - name: Provision Darwin
+        run: bash .github/workflows/provision-darwin.sh
```
```suggestion
      - name: Provision Darwin
        env:
          DFX_VERSION: 0.15.1
        run: bash .github/workflows/provision-darwin.sh
```

sesi200 commented on .github/workflows/motoko-icrc2-swap-example.yml:
```diff
@@ -0,0 +1,41 @@
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: actions/checkout@v1
+      - name: Provision Linux
+        run: bash .github/workflows/provision-linux.sh
```
```suggestion
      - name: Provision Linux
        env:
          DFX_VERSION: 0.15.1
        run: bash .github/workflows/provision-linux.sh
```

0xAegir commented:
Oh, seems like this is still running on DFX 0.12.0 (https://github.com/dfinity/examples/actions/runs/6783203421/job/18437037110?pr=621#step:3:210)... :/

sesi200 reviewed (changes requested):
Still haven't updated the default dfx version. It is missing the `--storage-mode` flag. If we request the most recent one it should work fine

sesi200 commented:
Btw there's a button like 'give maintainers write access to branch'. If you're willing to give me access this could be a bit easier

sesi200 commented:
Still failing with node errors, but on macOS this time...

0xAegir commented:
Yeah, weirdly github wasn't showing that option on this PR. I gave you write access to the repo anyway. 🙏

0xAegir commented:
@sesi200 should we pull the CI component out into a separate PR, so this doesn't stall?

sesi200 commented:
I'm not a fan of merging untested code, but since linux passes we could only postpone macOS for a separate PR

0xAegir commented:
Ok, I removed the macos CI test from this PR, and can do it as a follow-up.
